File: treebolicGlue/src/main/java/treebolic/glue/component/Container.java
/*
 * Copyright (c) 2019. Bernard Bou <1313ou@gmail.com>
 */

package treebolic.glue.component;

import android.content.Context;
import android.graphics.Point;
import android.graphics.drawable.ColorDrawable;
import android.graphics.drawable.Drawable;
import android.view.Gravity;
import android.view.View;
import android.view.ViewGroup;
import android.widget.LinearLayout;

import org.treebolic.glue.R;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

/**
 * Container
 * API class
 *
 * @author Bernard Bou
 */
public class Container extends LinearLayout implements Component, treebolic.glue.iface.component.Container<Component>
{
	static private float splitterPositionPercent = .80f;

	private final boolean isHorizontal;

	/**
	 * View
	 */
	@Nullable
	private View view;

	/**
	 * Toolbar
	 */
	@Nullable
	private View toolbar;

	/**
	 * Statusbar
	 */
	@Nullable
	private View statusbar;

	/**
	 * Constructor
	 *
	 * @param context context
	 */
	@SuppressWarnings("WeakerAccess")
	protected Container(@NonNull final Context context)
	{
		super(context);

		// determine orientation
		@NonNull final Point size = Utils.screenSize(context);
		this.isHorizontal = size.x >= size.y;
		setOrientation(this.isHorizontal ? LinearLayout.HORIZONTAL : LinearLayout.VERTICAL);

		// other
		setGravity(Gravity.CENTER);
		setWeightSum(1F);
	}

	/**
	 * Constructor
	 * API
	 *
	 * @param handle context
	 */
	public Container(final Object handle)
	{
		this((Context) handle);
	}

	@SuppressWarnings("WeakerAccess")
	@Override
	public void addComponent(final Component component, final int position)
	{
		final View viewToAdd = (View) component;

		switch (position)
		{
			case PANE:
				@NonNull final LayoutParams params = new LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.MATCH_PARENT, 1.F);
				addView(viewToAdd, params);
				break;

			// delayed add (until validation)
			case VIEW:
				this.view = viewToAdd;
				break;

			// delayed add (until validation)
			case TOOLBAR:
				this.toolbar = viewToAdd;
				break;

			// delayed add (until validation)
			case STATUSBAR:
				this.statusbar = viewToAdd;
				break;

			default:
				break;
		}
	}

	@SuppressWarnings("WeakerAccess")
	@Override
	public void removeAll()
	{
		removeAllViews();
		this.view = null;
		this.toolbar = null;
		this.statusbar = null;
	}

	@SuppressWarnings("WeakerAccess")
	@Override
	public void validate()
	{
		@Nullable @SuppressWarnings("UnusedAssignment") LayoutParams params = null;
		if (this.toolbar != null)
		{
			params = new LayoutParams(this.isHorizontal ? ViewGroup.LayoutParams.WRAP_CONTENT : ViewGroup.LayoutParams.MATCH_PARENT, this.isHorizontal ? ViewGroup.LayoutParams.MATCH_PARENT : ViewGroup.LayoutParams.WRAP_CONTENT, 0.F);
			addView(this.toolbar, params);
		}
		@NonNull ViewGroup layout = this;
		if (this.view != null && this.statusbar != null)
		{
			@NonNull final SplitPaneLayout splitLayout = new SplitPaneLayout(getContext());
			splitLayout.setOrientation(this.isHorizontal ? 0 : 1);
			splitLayout.setSplitterPositionPercent(Container.splitterPositionPercent);
			splitLayout.setSplitterMovable(true);
			splitLayout.setSplitterDrawable(getSplitterDrawable(false));
			splitLayout.setSplitterDraggingDrawable(getSplitterDrawable(true));

			this.addView(splitLayout);
			layout = splitLayout;
		}
		if (this.view != null)
		{
			params = new LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.MATCH_PARENT, 1.F);
			layout.addView(this.view, params);
		}
		if (this.statusbar != null)
		{
			params = new LayoutParams(this.isHorizontal ? ViewGroup.LayoutParams.WRAP_CONTENT : ViewGroup.LayoutParams.MATCH_PARENT, this.isHorizontal ? ViewGroup.LayoutParams.MATCH_PARENT : ViewGroup.LayoutParams.WRAP_CONTENT, 0.F);
			layout.addView(this.statusbar, params);
		}

		invalidate();
	}

	/**
	 * Get splitter position fraction (first panel/whole)
	 *
	 * @return splitterPositionPercent splitter position fraction [0,1]
	 */
	public static float getSplitterPositionPercent()
	{
		return Container.splitterPositionPercent;
	}

	/**
	 * Set splitter position fraction (first panel/whole)
	 *
	 * @param splitterPositionPercent splitter position fraction [0,1]
	 */
	public static void setSplitterPositionPercent(float splitterPositionPercent)
	{
		Container.splitterPositionPercent = splitterPositionPercent;
	}

	/**
	 * Get splitter drawable
	 *
	 * @param dragging dragging splitter
	 * @return drawable
	 */
	@NonNull
	private Drawable getSplitterDrawable(final boolean dragging)
	{
		@NonNull final int[] colors = Utils.fetchColors(getContext(), R.attr.treebolic_splitbar_color, R.attr.treebolic_splitbar_drag_color);
		return new ColorDrawable(colors[dragging ? 1 : 0]);
	}

	/*
	private Drawable getSplitterDrawable0(final boolean dragging)
	{
		final Resources res = getResources();
		try
		{
			return Drawable.createFromXml(res, res.getXml(dragging ? //
					this.isHorizontal ? //
							R.xml.splitter_bg_move_v : //
							R.xml.splitter_bg_move_h //
					: this.isHorizontal ? //
					R.xml.splitter_bg_v : //
					R.xml.splitter_bg_h));
		}
		catch (final Exception ex)
		{
			Log.e("TAG", "Exception loading drawable");
		}
		return null;
	}
	*/
}

File: treebolicGlue/src/main/java/treebolic/glue/component/Container.kt
/*
 * Copyright (c) 2019-2023. Bernard Bou
 */
package treebolic.glue.component

import android.content.Context
import android.graphics.drawable.ColorDrawable
import android.graphics.drawable.Drawable
import android.view.Gravity
import android.view.View
import android.view.ViewGroup
import android.widget.LinearLayout
import org.treebolic.glue.R
import treebolic.glue.iface.component.Container

/**
 * Container
 * API class
 *
 * @param handle context
 *
 * @author Bernard Bou
 */
open class Container(handle: Any?) : LinearLayout(handle as Context?), Component, Container<Component?> {

    private val isHorizontal: Boolean

    /**
     * View
     */
    private var view: View? = null

    /**
     * Toolbar
     */
    private var toolbar: View? = null

    /**
     * Statusbar
     */
    private var statusbar: View? = null

    /**
     * Constructor
     */
    init {
        // determine orientation
        val size = Utils.screenSize(context)
        isHorizontal = size.x >= size.y
        orientation = if (isHorizontal) HORIZONTAL else VERTICAL

        // other
        gravity = Gravity.CENTER
        weightSum = 1f
    }

    override fun addComponent(component: Component?, position: Int) {
        val viewToAdd = component as View?

        when (position) {
            Container.PANE -> {
                val params = LayoutParams(LayoutParams.MATCH_PARENT, LayoutParams.MATCH_PARENT, 1f)
                addView(viewToAdd, params)
            }

            Container.VIEW -> view = viewToAdd
            Container.TOOLBAR -> toolbar = viewToAdd
            Container.STATUSBAR -> statusbar = viewToAdd
            else -> {}
        }
    }

    override fun removeAll() {
        removeAllViews()
        view = null
        toolbar = null
        statusbar = null
    }

    override fun validate() {
        var params: LayoutParams?
        if (toolbar != null) {
            params = LayoutParams(if (isHorizontal) LayoutParams.WRAP_CONTENT else LayoutParams.MATCH_PARENT, if (isHorizontal) LayoutParams.MATCH_PARENT else LayoutParams.WRAP_CONTENT, 0f)
            addView(toolbar, params)
        }
        var layout: ViewGroup = this
        if (view != null && statusbar != null) {
            val splitLayout = SplitPaneLayout(context)
            splitLayout.orientation = if (isHorizontal) 0 else 1
            splitLayout.isSplitterMovable = true
            splitLayout.splitterDrawable = getSplitterDrawable(false)
            splitLayout.splitterDraggingDrawable = getSplitterDrawable(true)
            splitLayout.positionSplitterPercent(splitterPositionPercent)

            addView(splitLayout)
            layout = splitLayout
        }
        if (view != null) {
            params = LayoutParams(LayoutParams.MATCH_PARENT, LayoutParams.MATCH_PARENT, 1f)
            layout.addView(view, params)
        }
        if (statusbar != null) {
            params = LayoutParams(if (isHorizontal) LayoutParams.WRAP_CONTENT else LayoutParams.MATCH_PARENT, if (isHorizontal) LayoutParams.MATCH_PARENT else LayoutParams.WRAP_CONTENT, 0f)
            layout.addView(statusbar, params)
        }

        invalidate()
    }

    /**
     * Get splitter drawable
     *
     * @param dragging dragging splitter
     * @return drawable
     */
    private fun getSplitterDrawable(dragging: Boolean): Drawable {
        val colors = Utils.fetchColors(context, R.attr.treebolic_splitbar_color, R.attr.treebolic_splitbar_drag_color)
        return ColorDrawable(colors[if (dragging) 1 else 0])
    }

    companion object {

        /**
         * Splitter position fraction (first panel/whole)
         */
        var splitterPositionPercent: Float = .80f
    }
}
